zipline/utils/events.py

```diff
@@ -522,19 +502,45 @@ def get_last_trading_day_of_week(dt, env):
         dt = env.next_trading_day(dt)
         # Traverse forward until we hit a week border, then jump back to the
         # previous trading day.
-        while dt.date().weekday() > prev.date().weekday():
+        while dt and dt.date().weekday() > prev.date().weekday():
             prev = dt
             dt = env.next_trading_day(dt)
 
         if env.is_trading_day(prev):
-            return prev.date()
+            return prev
         else:
-            return env.previous_trading_day(prev).date()
+            return env.previous_trading_day(prev)
 
     date_func = get_last_trading_day_of_week
 
 
-class NthTradingDayOfMonth(StatelessRule):
+class TradingDayOfMonthRule(six.with_metaclass(ABCMeta, StatelessRule)):
+
+    def should_trigger(self, dt, env):
+        return self.get_trigger_day_of_month(dt, env) == dt.date()
+
+    @abstractmethod
+    def date_func(self, dt, env):
+        raise NotImplementedError
+
+    def get_trigger_day_of_month(self, dt, env):
+        if self.month == dt.month:
+            # We already computed the day for this month.
+            return self.day
+
+        self.day = self.date_func(dt, env)
+        if self.td_delta and self.day:
+            try:
+                self.day = env.add_trading_days(self.td_delta, self.day)
+            except NoFurtherDataError:
+                self.day = None
+            if self.day:
+                self.day = self.day.date()
```

Comment:
Is it intentional that here `self.day` is assigned a `date`, but above it's assigned a `datetime`?